File: https_server.py
"""HTTPS 预览服务器 - 支持 Service Worker 通知测试"""
import http.server, ssl, socket, os, sys

PORT = 8766
DIR = os.path.dirname(os.path.abspath(__file__))

class Handler(http.server.SimpleHTTPRequestHandler):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, directory=DIR, **kwargs)
    
    def end_headers(self):
        self.send_header('Service-Worker-Allowed', '/')
        self.send_header('Cache-Control', 'no-cache, no-store, must-revalidate')
        self.send_header('Pragma', 'no-cache')
        self.send_header('Expires', '0')
        super().end_headers()

    def log_message(self, format, *args):
        pass

httpd = http.server.HTTPServer(('0.0.0.0', PORT), Handler)
ctx = ssl.SSLContext(ssl.PROTOCOL_TLS_SERVER)
ctx.load_cert_chain(os.path.join(DIR, 'cert.pem'), os.path.join(DIR, 'key.pem'))
httpd.socket = ctx.wrap_socket(httpd.socket, server_side=True)

# 守护进程化：脱离当前 shell 会话
pid = os.fork()
if pid > 0:
    os._exit(0)
os.setsid()

# 获取局域网 IP（供手机访问）
local_ip = '127.0.0.1'
try:
    import subprocess
    r = subprocess.run(['ipconfig', 'getifaddr', 'en0'], capture_output=True, text=True)
    ip = r.stdout.strip()
    if ip:
        local_ip = ip
    else:
        r = subprocess.run(['ipconfig', 'getifaddr', 'en1'], capture_output=True, text=True)
        ip = r.stdout.strip()
        if ip:
            local_ip = ip
except:
    pass

print(f"\n  ✅ 预览服务器已启动")
print(f"  ─────────────────────────────────────")
print(f"  📱 手机访问 (HTTPS):")
print(f"     https://{local_ip}:{PORT}/")
print(f"  ─────────────────────────────────────")
print(f"  ⚠️  首次访问需接受证书警告")
print(f"     Safari → 点击「显示详情」→「访问此网站」")
print(f"  ─────────────────────────────────────\n")

try:
    httpd.serve_forever()
except KeyboardInterrupt:
    httpd.shutdown()

File: https_server.js
const https = require('https');
const fs = require('fs');
const path = require('path');
const url = require('url');

const PORT = 8766;
const DIR = __dirname;

const options = {
  key: fs.readFileSync(path.join(DIR, 'key.pem')),
  cert: fs.readFileSync(path.join(DIR, 'cert.pem'))
};

const MIME = {
  '.html': 'text/html; charset=utf-8',
  '.js': 'application/javascript; charset=utf-8',
  '.json': 'application/json',
  '.png': 'image/png',
  '.ico': 'image/x-icon',
  '.svg': 'image/svg+xml'
};

const server = https.createServer(options, (req, res) => {
  let filePath = path.join(DIR, req.url === '/' ? 'index.html' : req.url);
  filePath = path.normalize(filePath).replace(/\/$/, '/index.html');

  if (!filePath.startsWith(DIR)) {
    res.writeHead(403);
    res.end('Forbidden');
    return;
  }

  const ext = path.extname(filePath);
  const contentType = MIME[ext] || 'application/octet-stream';

  // PWA headers
  res.setHeader('Service-Worker-Allowed', '/');
  res.setHeader('Cache-Control', 'no-cache, no-store, must-revalidate');

  fs.readFile(filePath, (err, data) => {
    if (err) {
      res.writeHead(404);
      res.end('Not Found');
      return;
    }
    res.writeHead(200, { 'Content-Type': contentType });
    res.end(data);
  });
});

const os = require('os');
const interfaces = os.networkInterfaces();
let localIP = '127.0.0.1';
for (const name of Object.keys(interfaces)) {
  for (const iface of interfaces[name]) {
    if (iface.family === 'IPv4' && !iface.internal) {
      localIP = iface.address;
      break;
    }
  }
}

server.listen(PORT, '0.0.0.0', () => {
  console.log('\n  ✅ HTTPS 预览服务器已启动');
  console.log('  ─────────────────────────────────────');
  console.log(`  📱 手机访问:`);
  console.log(`     https://${localIP}:${PORT}/`);
  console.log('  ─────────────────────────────────────');
  console.log('  ⚠️  首次访问需接受证书安全警告');
  console.log('  ─────────────────────────────────────\n');
});
